src/lib.rs

```
@@ -0,0 +1,277 @@
+                .sum::<G1Projective>();
+
+        // Compute the final signature component `A = d_inv * B`
+        let a: G1Projective = b * d_inv;
+
+        Ok(Self { a, e })
+    }
+
+    /// Verifies a BBS MAC by recomputing B and checking that `A == (x + e)^{-1} * B`.
+    ///
+    /// # Parameters
+    /// - x: the secret key scalar
+    /// - messages: the messages to be verified
+    /// - mac: the MAC to be verified
+    /// - bbs_params: the system parameters
+    ///
+    /// # Returns
+    /// - `Ok(())`: the MAC is valid
+    /// - `Err(BbsMacError::VerificationInverseFailed)`: could not invert (x + e)
+    /// - `Err(BbsMacError::InputLengthMismatch)`: number of messages does not match number of generators
+    /// - `Err(BbsMacError::VerificationFailed)`: computed value does not match the provided MAC
+    ///
+    /// # Example
+    /// ```rust
+    /// use bbs_mac_vs_sig::{BbsParams, BbsMac};
+    /// use bls12_381::Scalar;
+    /// use ff::Field;
+    ///
+    /// let mut rng = rand::thread_rng();
+    /// let num_messages = 5;
+    /// let params = BbsParams::new(num_messages, &mut rng);
+    /// let x = Scalar::random(&mut rng);
+    ///
+    /// let mut messages = Vec::with_capacity(num_messages);
+    /// for _ in 0..num_messages {
+    ///     messages.push(Scalar::random(&mut rng));
+    /// }
+    ///
+    /// let mac = BbsMac::sign(&x, &messages, &params, &mut rng).unwrap();
+    /// BbsMac::verify(&x, &messages, &mac, &params).unwrap();
+    /// ```
+    pub fn verify(
+        x: &Scalar,
+        messages: &[Scalar],
+        mac: &BbsMac,
+        bbs_params: &BbsParams<G1Projective>,
+    ) -> Result<(), BbsMacError> {
+        let inv = (x + mac.e).invert();
+        if bool::from(inv.is_none()) {
+            return Err(BbsMacError::VerificationInverseFailed);
+        }
+        if messages.len() != bbs_params.h_generators.len() {
+            return Err(BbsMacError::InputLengthMismatch);
+        }
```

Fail fast: It's good practice to do as much input validation as possible before doing the more expenssive operations. In particular, I would check this before trying to do `.invert()`.